bson/primitive/decimal.go

```diff
@@ -139,169 +220,117 @@ func dErr(s string) (Decimal128, error) {
-	orig := s
-	if s == "" {
-		return dErr(orig)
-	}
-	neg := s[0] == '-'
-	if neg || s[0] == '+' {
-		s = s[1:]
-	}
-
-	if (len(s) == 3 || len(s) == 8) && (s[0] == 'N' || s[0] == 'n' || s[0] == 'I' || s[0] == 'i') {
+	matches := regDecimal128.FindStringSubmatch(s)
+	if len(matches) == 0 {
 		if s == "NaN" || s == "nan" || strings.EqualFold(s, "nan") {
 			return dNaN, nil
 		}
 		if s == "Inf" || s == "inf" || strings.EqualFold(s, "inf") || strings.EqualFold(s, "infinity") {
-			if neg {
-				return dNegInf, nil
-			}
 			return dPosInf, nil
 		}
-		return dErr(orig)
+		if s == "-Inf" || s == "-inf" || strings.EqualFold(s, "-inf") || strings.EqualFold(s, "-infinity") {
+			return dNegInf, nil
+		}
+		return dErr(s)
 	}
 
-	var h, l uint64
-	var e int
-
-	var add, ovr uint32
-	var mul uint32 = 1
-	var dot = -1
-	var digits = 0
-	var i = 0
-	for i < len(s) {
-		c := s[i]
-		if mul == 1e9 {
-			h, l, ovr = muladd(h, l, mul, add)
-			mul, add = 1, 0
-			if ovr > 0 || h&((1<<15-1)<<49) > 0 {
-				return dErr(orig)
-			}
-		}
-		if c >= '0' && c <= '9' {
-			i++
-			if c > '0' || digits > 0 {
-				digits++
-			}
-			if digits > 34 {
-				if c == '0' {
-					// Exact rounding.
-					e++
-					continue
-				}
-				return dErr(orig)
-			}
-			mul *= 10
-			add *= 10
-			add += uint32(c - '0')
-			continue
-		}
-		if c == '.' {
-			i++
-			if dot >= 0 || i == 1 && len(s) == 1 {
-				return dErr(orig)
-			}
-			if i == len(s) {
-				break
-			}
-			if s[i] < '0' || s[i] > '9' || e > 0 {
-				return dErr(orig)
-			}
-			dot = i
-			continue
+	ip := matches[1]
+	dp := matches[2]
+	ep := matches[3]
+
+	var err error
+	e := 0
+	if ep != "" {
+		e, err = strconv.Atoi(ep)
+		if err != nil {
+			return dErr(s)
 		}
-		break
 	}
-	if i == 0 {
-		return dErr(orig)
+	if dp != "" {
+		e -= len(dp)
 	}
-	if mul > 1 {
-		h, l, ovr = muladd(h, l, mul, add)
-		if ovr > 0 || h&((1<<15-1)<<49) > 0 {
-			return dErr(orig)
-		}
+
+	if len(strings.Trim(ip+dp, "-0")) > 35 {
+		return dErr(s)
 	}
-	if dot >= 0 {
-		e += dot - i
+
+	bi, ok := new(big.Int).SetString(ip+dp, 10)
+	if !ok {
+		return dErr(s)
 	}
-	if i+1 < len(s) && (s[i] == 'E' || s[i] == 'e') {
-		i++
-		eneg := s[i] == '-'
-		if eneg || s[i] == '+' {
-			i++
-			if i == len(s) {
-				return dErr(orig)
-			}
-		}
-		n := 0
-		for i < len(s) && n < 1e4 {
-			c := s[i]
-			i++
-			if c < '0' || c > '9' {
-				return dErr(orig)
-			}
-			n *= 10
-			n += int(c - '0')
-		}
-		if eneg {
-			n = -n
+
+	d, ok := ParseDecimal128FromBigInt(bi, e)
+	if !ok {
+		return dErr(s)
+	}
+	return d, nil
+}
+
+var (
+	ten  = big.NewInt(10)
+	zero = new(big.Int)
+	maxS = new(big.Int).SetBytes([]byte{0x1, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff}) // 113 bits
+)
+
+// ParseDecimal128FromBigInt attempts to parse the given significand and exponent into a valid Decimal128 value.
+func ParseDecimal128FromBigInt(bi *big.Int, exp int) (Decimal128, bool) {
+	//copy
+	if bi == nil {
```

Comment:
Is it valid to have bi be nil?

Reply:
big.Int is similar to int is a number. I think number has default zero value is in line with the go style.